### PR TITLE
refactor(language-service): Construct proper template AST from HTML ast

### PR DESCRIPTION
**Pull request by kyliau**
The Template AST that corresponds to a given HTML AST is not always
complete, and often has to be reconstructed. This commit refactors the
code to make it easier to adapt to multiple cases.

This is a follow up to https://github.com/angular/angular/pull/34743

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by kyliau** [comment]
@ivanwonder Do you mind taking a look at this, please?

File: packages/language-service/src/completions.ts [hunk 1]
```diff
@@ -300,44 +296,40 @@ function attributeCompletionsForElement(
...
+function attributeValueCompletions(info: AstResult, htmlPath: HtmlAstPath): ng.CompletionEntry[] {
+  // Find the corresponding Template AST path.
+  const templatePath = findTemplateAstAt(info.templateAst, htmlPath.position);
+  const visitor = new ExpressionVisitor(info, htmlPath.position, () => {
+    const dinfo = diagnosticInfoFromTemplateInfo(info);
+    return getExpressionScope(dinfo, templatePath, false);
+  });
+  if (templatePath.tail instanceof AttrAst ||
+      templatePath.tail instanceof BoundElementPropertyAst ||
+      templatePath.tail instanceof BoundEventAst) {
```
[the remark this answers is not in the document]
Comment (ivanwonder) [reply packages/language-service/src/completions.ts]: > can you add `ReferenceAst` as well? I don't think it ever happens, but probably worth a check..

It maybe happens, when the user begin to input reference(`<div #refs="|"></div>`).

[the remark this answers is not in the document]
Comment (kyliau) [reply packages/language-service/src/completions.ts]: I added `ReferenceAst` as a precaution.
This way of transforming Html Ast to Template Ast is a consequence of the ViewEngine architecture.
With the switch to Ivy later, this transformation is no longer necessary. Ivy uses an architecture similar to TypeScript, which allows us to query the Ivy TypeChecker given a HTML node.
So I wouldn't worry too much about bigger debt down the road - this code will be obsolete very soon ...

File: packages/language-service/src/completions.ts [hunk 2]
```diff
@@ -300,43 +296,38 @@ function attributeCompletionsForElement(
...
-function referenceAttributeValueCompletions(
-    info: AstResult, position: number, attr: Attribute): ng.CompletionEntry[] {
-  const path = findTemplateAstAt(info.templateAst, position);
-  if (!path.tail) {
-    return [];
+/**
+ * Provide completions to the RHS of an attribute, which is of the form
+ * LHS="RHS". The template path is computed from the specified `info` whereas
+ * the context is determined from the specified `htmlPath`.
+ * @param info
```
[the remark this answers is not in the document]
Comment (kyliau) [reply packages/language-service/src/completions.ts]: done

File: packages/language-service/src/completions.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivanwonder) [reply packages/language-service/src/completions.ts]: The only exception I know is `ReferenceAst`, Maybe it's better to move the code to line 330 in the `else`.
But it's ok for me as you say. Look forward to the new Ivy.😍

**Review by ayazhafiz — APPROVED** [review]

**Comment by ayazhafiz** [comment]
@kyliau linter ):

**Pull request closed without merging**